File: Libs/TauMPC/src/mpc/circuit/Wire.java
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */

package mpc.circuit;

import mpc.finite_field_math.Zp;


public class Wire {
    private  Integer inputIndex;
    private Integer outputIndex;
    private   Gate sourceGate;
    private   Gate targetGate;
    private Zp constValue;

    public Wire() {
        inputIndex = null;
        outputIndex = null;
    }        
        

    public Wire(Zp constValue) {
        this.constValue = constValue;
        inputIndex = null;
        outputIndex = null;
    }        

    public Wire(Integer inputOrOutputIndex, boolean isInput) {
        if (isInput) {
            this.inputIndex = inputOrOutputIndex;
            outputIndex = null;
        } else {
            this.outputIndex = inputOrOutputIndex;
            inputIndex = null;
        }
    }
       
    public Wire(Integer inputIndex, Gate targetGate) {
        this.inputIndex = inputIndex;
        this.targetGate = targetGate;
        outputIndex = null;
        sourceGate = null;        
    }

    public Wire(Gate sourceGate, Integer outputIndex) {
        this.outputIndex = outputIndex;
        this.sourceGate = sourceGate;
        inputIndex = null;
        targetGate = null;
    }

    public Wire(Gate sourceGate, Gate targetGate) {
        this.sourceGate = sourceGate;
        this.targetGate = targetGate;
        inputIndex = null;
        outputIndex = null;
    }

    public Wire(Integer inputIndex, Integer outputIndex) {
        this.inputIndex = inputIndex;
        this.outputIndex = outputIndex;
        sourceGate = null;
        targetGate = null;
    }
    
    public boolean isValid(){
        return inputIndex != null || sourceGate != null && sourceGate.isReadyToCalculate();
    }

    public boolean isInput() {
        return inputIndex != null;
    }

    public boolean isOutput() {
        return outputIndex != null;
    }

    public Gate getTargetGate() {
        return targetGate;
    }

    public void setTargetGate(Gate targetGate) {
        assert outputIndex == null : "output wire should not have target gate";
        this.targetGate = targetGate;
    }

    public Gate getSourceGate() {
        return sourceGate;
    }

    public void setSourceGate(Gate sourceGate) {
        assert inputIndex == null : "input wire should not have source gate";
        this.sourceGate = sourceGate;
    }

    public Zp getConstValue() {
        return constValue;
    }

    public void setConstValue(Zp constValue) {
        this.constValue = constValue;
    }        

    public Integer getOutputIndex() {
        return outputIndex;
    }

    public void setOutputIndex(Integer outputIndex) {
        this.outputIndex = outputIndex;
    }

    public Integer getInputIndex() {
        return inputIndex;
    }

    public void setInputIndex(Integer inputIndex) {
        this.inputIndex = inputIndex;
    }        
    
    @Override
    public Wire clone(){
        Wire wire = new Wire(inputIndex, outputIndex);
        wire.setSourceGate(sourceGate);
        wire.setTargetGate(targetGate);     
        wire.setConstValue(constValue);
        if (wire.getSourceGate() != null){
            wire.getSourceGate().addOutputWire(wire);            
        }      
        return wire;                
    }
}

File: Libs/FiniteField/Circuits/Wire.cs
using System.Diagnostics;

namespace MpcLib.Common.FiniteField.Circuits
{
	public class Wire
	{
		public int InputIndex;
		public bool IsOutput;
		public readonly Zp ConstValue;
		private Gate sourceGate;
		private Gate targetGate;

		#region Constructors

		public Wire()
		{
			InputIndex = -1;
		}

		public Wire(Zp constValue)
			: this()
		{
			this.ConstValue = constValue;
		}

		public Wire(int inputIndex, bool isInput)
		{
			if (isInput)
				this.InputIndex = inputIndex;
			else
			{
				this.IsOutput = true;
				InputIndex = -1;
			}
		}

		public Wire(int inputIndex, Gate targetGate)
		{
			this.InputIndex = inputIndex;
			this.targetGate = targetGate;
		}

		public Wire(Gate sourceGate, bool isOutput)
		{
			this.IsOutput = isOutput;
			this.sourceGate = sourceGate;
			InputIndex = -1;
		}

		public Wire(Gate sourceGate, Gate targetGate)
			: this()
		{
			this.sourceGate = sourceGate;
			this.targetGate = targetGate;
		}

		public Wire(int inputIndex, bool isOutput, Zp constValue)
		{
			this.InputIndex = inputIndex;
			this.IsOutput = isOutput;
			ConstValue = constValue;
		}

		#endregion Constructors

		public virtual bool Valid
		{
			get
			{
				return InputIndex != -1 || sourceGate != null && sourceGate.ReadyToCalculate;
			}
		}

		public virtual bool IsInput
		{
			get
			{
				return InputIndex != -1;
			}
		}

		public virtual Gate TargetGate
		{
			get
			{
				return targetGate;
			}

			set
			{
				Debug.Assert(!IsOutput, "output wire should not have target gate");
				this.targetGate = value;
			}
		}

		public virtual Gate SourceGate
		{
			get
			{
				return sourceGate;
			}

			set
			{
				Debug.Assert(value == null || InputIndex == -1, "input wire should not have source gate");
				this.sourceGate = value;
			}
		}

		public Wire Clone()
		{
			var wire = new Wire(InputIndex, IsOutput, ConstValue);
			wire.SourceGate = sourceGate;
			wire.TargetGate = targetGate;
			if (wire.SourceGate != null)
				wire.SourceGate.AddOutputWire(wire);
			return wire;
		}
	}
}